src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,920 @@
+												   List *splitPointsForShard);
+static void CreateSplitIntervalsForShard(ShardInterval *sourceShard,
+										 List *splitPointsForShard,
+										 List **shardSplitChildrenIntervalList);
+static void BlockingShardSplit(SplitOperation splitOperation,
+							   ShardInterval *shardIntervalToSplit,
+							   List *shardSplitPointsList,
+							   List *workersForPlacementList);
+static void DoSplitCopy(WorkerNode *sourceShardNode,
+						List *sourceColocatedShardIntervalList,
+						List *shardGroupSplitIntervalListList,
+						List *workersForPlacementList);
+static StringInfo CreateSplitCopyCommand(ShardInterval *sourceShardSplitInterval,
+										 List *splitChildrenShardIntervalList,
+										 List *workersForPlacementList);
+static void InsertSplitChildrenShardMetadata(List *shardGroupSplitIntervalListList,
+											 List *workersForPlacementList);
+static void CreateForeignKeyConstraints(List *shardGroupSplitIntervalListList,
+										List *workersForPlacementList);
+static void TryDropSplitShardsOnFailure(List *shardGroupSplitIntervalListList,
+										List *workersForPlacementList);
+
+/* Customize error message strings based on operation type */
+static const char *const SplitOperationName[] =
+{
+	[SHARD_SPLIT_API] = "split",
+	[ISOLATE_TENANT_TO_NEW_SHARD] = "isolate",
+};
+static const char *const SplitTargetName[] =
+{
+	[SHARD_SPLIT_API] = "shard",
+	[ISOLATE_TENANT_TO_NEW_SHARD] = "tenant",
+};
+
+/* Function definitions */
+
+/*
+ * ErrorIfCannotSplitShard checks relation kind and invalid shards. It errors
+ * out if we are not able to split the given shard.
+ */
+void
+ErrorIfCannotSplitShard(SplitOperation splitOperation, ShardInterval *sourceShard)
+{
+	Oid relationId = sourceShard->relationId;
+	ListCell *colocatedTableCell = NULL;
+	ListCell *colocatedShardCell = NULL;
+
+	/* checks for table ownership and foreign tables */
+	List *colocatedTableList = ColocatedTableList(relationId);
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+
+		/* check that user has owner rights in all co-located tables */
+		EnsureTableOwner(colocatedTableId);
+
+		char relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			char *relationName = get_rel_name(colocatedTableId);
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot %s %s because \"%s\" is a "
+								   "foreign table",
+								   SplitOperationName[splitOperation],
+								   SplitTargetName[splitOperation],
+								   relationName),
+							errdetail("Splitting shards backed by foreign tables "
+									  "is not supported.")));
+		}
+
+		/*
+		 * At the moment, we do not support copying a shard if that shard's
+		 * relation is in a colocation group with a partitioned table or partition.
+		 */
+		if (PartitionedTable(colocatedTableId))
```

Comment:
We'll need to support partitioning. That's very commonly used.

Reply:
Not part of this first MVP PR. I will have a follow up PR and remove this check as part of that. I am tracking this item separately.

Reply:
As per our IM group discussion, resolving this. Will follow up in coming PR.